app/Http/Middleware/VerifyCsrfToken.php

```diff
@@ -23,38 +25,28 @@ class VerifyCsrfToken extends Middleware
 
 	/**
 	 * The goal of this function is to allow to bypass the CSRF token requirement
-	 * if an Authorization value is provided in the header and matches the apiKey.
-	 *
-	 * FIXME: Do we want to hash this API key ? Might actually be a good idea...
+	 * if an Authorization value is provided in the header and matches the
+	 * token of a user.
 	 *
 	 * @param Request $request
 	 * @param Closure $next
 	 *
 	 * @return mixed
 	 *
 	 * @throws TokenMismatchException
+	 * @throws QueryBuilderException
 	 */
 	public function handle($request, Closure $next): mixed
 	{
-		if ($request->is('api/*')) {
-			/**
-			 * default value is ''
-			 * we force it in case of the migration has not been done.
-			 */
-			$apiKey = Configs::getValueAsString('api_key');
-
-			/*
-			 * if apiKey is the empty string we directly return the parent handle.
-			 */
-			if ($apiKey === '') {
-				return parent::handle($request, $next);
-			}
+		$token = $request->header('Authorization');
+		if ($request->is('api/*') && is_string($token)) {
+			/** @var User|null $user */
+			$user = User::query()
+				->where('token', '=', hash('SHA512', $token))
+				->first();
+			if ($user instanceof User) {
+				Auth::loginUsingId($user->id);
```

Comment:
Authentication as the corresponding user is something new. I understand the intention of the code, but I am unsure whether we are good here.

Normally, the authenticated user is stored as part of the session. The session ID is persisted as a cookie between individual requests. Some other middelware re-reads the cookie value and starts the session (`\Illuminate\Session\Middleware\StartSession`) and still another one authenticates the user stored in that session (`\Illuminate\Session\Middleware\AuthenticateSession`). Here we interfere with this authentication. I don't know (yet) in which order these middelwares are executed of if there is even a deterministic order. I guess the middleware which is executed last and calls `Auth::login` "wins".

My question is: What happens if there is a session with a stored user inside it and the user authenticated via the session and the user authenticated via the API key are out-of-sync?
This would lead to some kind of unintentional re-authentication as a different user. Normally, Laravel requires us to call `Session::flush` and therewith destroy the session for intentional changes in the authenticated user. Obviously, that is nothing what we can or should do here, because we don't want to destroy the session.

However, I am wondering whether we need to take additional steps of precaution here and in `\Illuminate\Session\Middleware\AuthenticateSession`. Something like checking if there is a stored user in the session and a user associated to the API key, both user objects are referring to the same user and if not bail out.

The following somewhat "freaky" scenario is on my mind:

- A client starts a session and authenticates a some user A
- A client uses the same session ID and sends the API key for another user B

Of course, this should not happen for a well-behaving client, but you never know. Such crazy things do happen.

Reply:
Thanks. At least nothing bad seems to happen.

Still I would like to run this through the step debugger and have a closer look at the `AuthenticateSession` middleware, if I can spot anything which might look suspicious like a lurking bug. I will do so this weekend.

Reply:
I fixed the concerns I raised above in https://github.com/LycheeOrg/Lychee/pull/1368/commits/c5d2ff62ccfa44319f06b7ad228a2e4f8046d0a1.

It took me hours to find out how Laravel Authentication is supposed to work. The new guard should also handle all "bad" cases correctly. I haven't tried it yet, just coded. (It took longer than expected.) So chances are that there are some lurking bugs.

@qwerty287 You should play around with it, if this PR still does what you want it to do after the change.